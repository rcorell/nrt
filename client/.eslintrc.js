module.exports = {
    env: {
        browser: true,
        commonjs: true,
        es6: true,
        jest: true,
        node: true
    },
    extends: ['react-app', 'eslint:recommended', 'plugin:@typescript-eslint/recommended', 'plugin:react/recommended', 'prettier'],
    parser: '@typescript-eslint/parser',
    parserOptions: {
        ecmaFeatures: {
            jsx: true
        }
    },
    plugins: ['@typescript-eslint', 'import', 'prettier'],
    root: true,
    rules: {
        '@typescript-eslint/ban-types': 'off',
        '@typescript-eslint/explicit-function-return-type': 'off',
        '@typescript-eslint/explicit-module-boundary-types': 'off',
        '@typescript-eslint/no-explicit-any': 'off',
        '@typescript-eslint/no-use-before-define': 'off',
        'import/order': [
            'error',
            {
                groups: ['builtin', 'external', 'internal'],
                'newlines-between': 'always',
                alphabetize: {
                    order: 'asc',
                    caseInsensitive: true
                }
            }
        ],
        'prettier/prettier': 2,
        'sort-imports': ['error', { ignoreCase: true, ignoreDeclarationSort: true }]
    },
    settings: {
        'import/resolver': {
            typescript: {
                project: './tsconfig.json'
            }
        },
        react: {
            version: "detect"
          }
    }
};
