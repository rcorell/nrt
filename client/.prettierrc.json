{
  "arrowParens": "avoid",
  "bracketSpacing": true,
  "printWidth": 116,
  "semi": true,
  "singleQuote": true,
  "tabWidth": 4,
  "trailingComma": "none"
}
