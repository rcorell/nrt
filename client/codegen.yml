overwrite: true
schema: "http://localhost:4000"
documents: "src/**/*.graphql"
generates:
  src/api/__generated__/types.ts:
    hooks:
      afterOneFileWrite:
        - prettier --write
        - eslint --fix
    plugins:
      - "typescript"
      - "typescript-operations"
      - "typescript-react-apollo"
  ./graphql.schema.json:
    plugins:
      - "introspection"



