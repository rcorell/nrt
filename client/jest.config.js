module.exports = {
    clearMocks: true,
    collectCoverageFrom: ['<rootDir>/src/**/*.{js,jsx,ts,tsx}', '!<rootDir>/src/**/*.d.ts'],
    coveragePathIgnorePatterns: ['types.ts'],
    coverageThreshold: {
        global: {
            branches: 100,
            functions: 100,
            lines: 100,
            statements: 0
        }
    },
    globals: {
        'ts-jest': {
            diagnostics: 'pretty'
        }
    },
    moduleFileExtensions: ['js', 'ts', 'tsx'],
    modulePaths: ['<rootDir>'],
    preset: 'ts-jest',
    roots: ['<rootDir>/src/', '<rootDir>/tests/'],
    setupFiles: ['<rootDir>/tests/setupTests.ts'],
    setupFilesAfterEnv: ['<rootDir>/tests/setupTestFrameworkAdditions.tsx'],
    snapshotSerializers: ['enzyme-to-json/serializer'],
    testEnvironment: 'node',
    testMatch: ['<rootDir>/tests/**/?(*.)(spec|test).(j|t)s?(x)'],
    testURL: 'http://localhost',
    transform: {
        '^.+\\.css$': '<rootDir>/tests/cssTransform.js',
        '^.+\\.tsx?$': 'ts-jest'
    },
    transformIgnorePatterns: ['[/\\\\]node_modules[/\\\\].+\\.(js|jsx|mjs|ts|tsx)$']
};
