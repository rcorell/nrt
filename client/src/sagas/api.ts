import { SagaIterator } from 'redux-saga';
import { all, takeLatest } from 'redux-saga/effects';
import { ActionType } from 'typesafe-actions';

import * as signupActions from '../redux/signup/actions';

export function* watcher(): SagaIterator<void> {
    // yield call(_checkAuthenticated);

    yield all([
        takeLatest(signupActions.signup, _signup),
    ]);
}

export function* _signup(action: ActionType<typeof signupActions.signup>): SagaIterator<void> {
	const { email, name, password } = action.payload;

	console.log(email, name, password);
}
