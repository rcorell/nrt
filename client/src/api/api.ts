import { ApolloClient, ApolloLink, gql, HttpLink, InMemoryCache } from '@apollo/client';

const AUTHENTICATION_TOKEN_NAME = 'token';

const httpLink = new HttpLink({ uri: 'http://localhost:4000' });

const authLink = new ApolloLink((operation, forward) => {
    const token = localStorage.getItem(AUTHENTICATION_TOKEN_NAME);

    operation.setContext({
        headers: {
            authorization: token ? `Bearer ${token}` : ''
        }
    });

    return forward(operation);
});

export const apolloClient = new ApolloClient({
    link: authLink.concat(httpLink),
    cache: new InMemoryCache()
});

export const isAuthenticated = () => {
    return Boolean(localStorage.getItem(AUTHENTICATION_TOKEN_NAME));
};

export const setAuthenticationToken = (token: string) => {
    localStorage.setItem(AUTHENTICATION_TOKEN_NAME, token);
};

export const fetchTopicsQueryString = gql`
    query FetchTopicsQuery {
        topics {
            title
            description
        }
    }
`;

export const CreateGroupMutation = gql`
    mutation CreateGroupMutation($name: String!, $description: String) {
        createGroup(name: $name, description: $description) {
            id
            createdAt
        }
    }
`;

export const SignupMutation = gql`
    mutation SignupMutation($email: String!, $name: String!, $password: String!) {
        signup(email: $email, name: $name, password: $password) {
            token
        }
    }
`;

export const loginMutationString = gql`
    mutation LoginMutation($email: String!, $password: String!) {
        login(email: $email, password: $password) {
            token
        }
    }
`;

export const CreateTopicMutation = gql`
    mutation CreateTopicMutation($title: String!, $description: String) {
        createTopic(title: $title, description: $description) {
            id
            createdAt
        }
    }
`;
