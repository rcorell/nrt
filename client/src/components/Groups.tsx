import React from 'react';
import * as _ from 'lodash';
import { useMutation, useQuery } from '@apollo/client';
import MaterialTable from 'material-table';

import { fetchGroupsQueryString, fetchUserQueryString, joinGroupMutationString } from 'src/api/api';
import { FetchGroupsQuery } from 'src/api/__generated__/FetchGroupsQuery';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { JoinGroupMutation, JoinGroupMutationVariables } from 'src/api/__generated__/JoinGroupMutation';
import { setBrowserTitle } from 'src/utils';

// const renderJoinColumn = (rowData: any) => {
//     if (userGroups.contains(rowData.name)) {
//         return <p>Joined</p>;
//     }

//     return <p>Join</p>;
// };

const COLUMN_DEFINITIONS = [
    { title: 'Name', field: 'name' },
    { title: 'Description', field: 'description' },
    {
        field: 'join',
        title: ''
    }
];

export const Groups: React.FC = () => {
    setBrowserTitle('Groups');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);
    const allGroups = useQuery<FetchGroupsQuery>(fetchGroupsQueryString);

    const [joinGroup] = useMutation<JoinGroupMutation, JoinGroupMutationVariables>(joinGroupMutationString, {
        onCompleted: () => {
            // console.log('woo');
            userQuery.refetch();
        },
        onError: () => {
            // https://github.com/apollographql/apollo-client/issues/7167
        }
    });

    if (allGroups.loading || userQuery.loading) {
        return <div>Loading...</div>;
    }

    if (allGroups.error || userQuery.error) {
        return <div>Error: {JSON.stringify(allGroups.error) + JSON.stringify(userQuery.error)}</div>;
    }

    const userGroupIds = userQuery!.data!.user.groups.map((group) => {
        return group!.id;
    });

    const groups = _.cloneDeep(allGroups.data!.groups).map((group) => {
        return {
            ...group,
            join: userGroupIds.includes(group.id) ? (
                'Joined'
            ) : (
                <button onClick={() => joinGroup({ variables: { groupId: Number(group.id) } })}>Join!</button>
            )
        };
    });

    return <MaterialTable title="Groups" columns={COLUMN_DEFINITIONS} data={groups} />;
};
