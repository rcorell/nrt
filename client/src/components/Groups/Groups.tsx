import _ from 'lodash';
import MaterialTable from 'material-table';
import React from 'react';

import { LOADING_TEXT } from 'src/components/shared';
import { setBrowserTitle } from 'src/utils';

import { useGroups, useJoinGroup, useUser } from './Groups.hook';

const COLUMN_DEFINITIONS = [
    { field: 'name', title: 'Name' },
    { field: 'description', title: 'Description' },
    {
        field: 'join',
        title: ''
    }
];

export const Groups: React.FC = () => {
    setBrowserTitle('Groups');

    const userHook = useUser();
    const groupsHook = useGroups();

    const joinGroupHook = useJoinGroup();

    if (groupsHook.loading || userHook.loading || joinGroupHook.loading) {
        return <div>{LOADING_TEXT}</div>;
    }
    if (groupsHook.error || userHook.error || joinGroupHook.error) {
        return (
            <div>
                {JSON.stringify(groupsHook.error)}
                <br />
                {JSON.stringify(userHook.error)}
                <br />
                {JSON.stringify(joinGroupHook.error)}
            </div>
        );
    }

    const userGroupIds = userHook.user!.groups.map((group) => {
        return group!.id;
    });

    const groups = _.cloneDeep(groupsHook!.allGroups!).map((group) => {
        return {
            ...group,
            join: userGroupIds.includes(group.id) ? (
                'Joined'
            ) : (
                <button onClick={() => joinGroupHook.joinGroup(group.id)}>Join!</button>
            )
        };
    });

    return (
        <MaterialTable
            title="Groups"
            columns={COLUMN_DEFINITIONS}
            data={groups}
            options={{
                pageSize: 20,
                pageSizeOptions: [20, 50, 100]
            }}
        />
    );
};
