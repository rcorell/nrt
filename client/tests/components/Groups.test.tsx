import { MockedResponse } from '@apollo/client/testing';
import { fireEvent, screen, waitForElementToBeRemoved } from '@testing-library/react';

import { fetchGroupsQueryString, fetchUserQueryString, joinGroupMutationString } from 'src/api/api';
import { Groups } from 'src/components/Groups';

import { oneTick, renderComponent, renderComponentWithMocks } from 'tests/testHelpers';

describe('Groups', () => {
    const LOADING_TEXT = 'Loading...';

    describe('snapshots', () => {
        it('loading', () => {
            expect(renderComponent(Groups).container).toMatchSnapshot();
        });

        it('network error', async () => {
            const networkError = new Error('Network-error');

            const { container } = renderComponent(Groups, fetchGroupsQueryString, {}, { error: networkError });

            await waitForElementToBeRemoved(() => screen.getByText(LOADING_TEXT));

            expect(screen.queryByText('No results')).toBeNull();

            expect(container).toMatchSnapshot();
        });
    });

    describe('success', () => {
        it('display one group', async () => {
            const mocks: MockedResponse[] = [
                {
                    request: {
                        query: fetchGroupsQueryString,
                        variables: {}
                    },
                    result: {
                        data: {
                            groups: [
                                {
                                    __typename: 'Group',
                                    id: '42',
                                    description: 'test-description',
                                    name: 'test-name'
                                }
                            ]
                        }
                    }
                },
                {
                    request: {
                        query: fetchUserQueryString,
                        variables: {}
                    },
                    result: {
                        data: {
                            user: {
                                id: '11',
                                groups: [],
                                topics: []
                            }
                        }
                    }
                }
            ];
            renderComponentWithMocks(Groups, mocks);

            await waitForElementToBeRemoved(() => screen.getByText(LOADING_TEXT));

            expect(screen.queryByText('test-description')).toBeInTheDocument();
            expect(screen.queryByText('test-name')).toBeInTheDocument();
        });

        it('join group', async () => {
            const mocks: MockedResponse[] = [
                {
                    request: {
                        query: fetchGroupsQueryString,
                        variables: {}
                    },
                    result: {
                        data: {
                            groups: [
                                {
                                    __typename: 'Group',
                                    id: '42',
                                    description: 'test-description',
                                    name: 'test-name'
                                }
                            ]
                        }
                    }
                },
                {
                    request: {
                        query: fetchUserQueryString,
                        variables: {}
                    },
                    result: {
                        data: {
                            user: {
                                id: '11',
                                groups: [],
                                topics: []
                            }
                        }
                    }
                },
                {
                    request: {
                        query: joinGroupMutationString,
                        variables: { groupId: '42' }
                    },
                    result: {
                        data: {
                            id: '42'
                        }
                    }
                },
                {
                    request: {
                        query: fetchUserQueryString,
                        variables: {}
                    },
                    result: {
                        data: {
                            user: {
                                id: '11',
                                groups: [{ id: '42', name: 'name', description: 'description' }],
                                topics: []
                            }
                        }
                    }
                }
            ];
            renderComponentWithMocks(Groups, mocks);

            await waitForElementToBeRemoved(() => screen.getByText(LOADING_TEXT));

            expect(screen.queryByText('test-description')).toBeInTheDocument();
            expect(screen.queryByText('test-name')).toBeInTheDocument();

            const joinButton = screen.getByText('Join!');
            fireEvent.click(joinButton);
            await oneTick();

            // expect(screen.queryByText('Joined')).toBeInTheDocument();
        });
    });
});
