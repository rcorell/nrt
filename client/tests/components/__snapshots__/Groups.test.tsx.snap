// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Groups snapshots loading 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`Groups snapshots network error 1`] = `
<div>
  <div>
    Error: 
    {"graphQLErrors":[],"networkError":{},"message":"Network-error"}{"graphQLErrors":[],"networkError":{},"message":"No more mocked responses for the query: query FetchUserQuery {\\n  user {\\n    groups {\\n      id\\n      description\\n      name\\n    }\\n    topics {\\n      id\\n      description\\n      title\\n    }\\n  }\\n}\\n, variables: {}"}
  </div>
</div>
`;
